plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerBehaviour.java

```diff
@@ -153,23 +158,64 @@ private IStatus newErrorStatus(String message) {
     return new Status(IStatus.ERROR, Activator.PLUGIN_ID, message);
   }
 
-  private int checkAndSetPort() throws CoreException {
-    port = getServer().getAttribute(SERVER_PORT_ATTRIBUTE_NAME, DEFAULT_SERVER_PORT);
-    if (port < 0 || port > 65535) {
+  private void checkAndSetPorts() throws CoreException {
+    serverPort = getServer().getAttribute(SERVER_PORT_ATTRIBUTE_NAME, DEFAULT_SERVER_PORT);
+
+    checkAndSetPorts(new PortProber() {
+      @Override
+      public boolean isPortInUse(int port) {
+        return org.eclipse.wst.server.core.util.SocketUtil.isPortInUse(port);
+      }
+
+      @Override
+      public List<Integer> findFreePorts(int portCounts) {
+        return SocketUtil.findFreePorts(portCounts);
+      }
+    });
+  }
+
+  @VisibleForTesting
+  public interface PortProber {
+    boolean isPortInUse(int port);
+    List<Integer> findFreePorts(int portCounts);
+  }
+
+  @VisibleForTesting
+  void checkAndSetPorts(PortProber portProber)
+      throws CoreException {
+    if (serverPort < 0 || serverPort > 65535) {
       throw new CoreException(newErrorStatus("Port must be between 0 and 65535."));
     }
+    if (serverPort != 0 && portProber.isPortInUse(serverPort)) {
+      throw new CoreException(newErrorStatus("Port " + serverPort + " is in use."));
+    }
+
+    if (portProber.isPortInUse(adminPort)) {
+      adminPort = 0;
+    }
 
-    if (port == 0) {
-      port = SocketUtil.findFreePort();
-      if (port == -1) {
+    if (serverPort == 0 || adminPort == 0) {
+      LinkedList<Integer> ports = null;
```

Comment:
Actually, I'm deliberately transforming `List` to `LinkedList` to use it as a queue. It's very convenient in this situation, since I can just pop the elements once or twice.